src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/PoolingDataSource.cs

```diff
@@ -45,19 +90,246 @@ internal sealed class PoolingDataSource : SqlDataSource
         {
             _connectionPoolGroupOptions = options;
             _connectionRateLimiter = connectionRateLimiter;
-            //TODO: other construction
+            _connectors = new SqlConnector[MaxPoolSize];
+
+            // We enforce Max Pool Size, so no need to to create a bounded channel (which is less efficient)
+            // On the consuming side, we have the multiplexing write loop but also non-multiplexing Rents
+            // On the producing side, we have connections being released back into the pool (both multiplexing and not)
+            var idleChannel = Channel.CreateUnbounded<SqlConnector?>();
+            _idleConnectorReader = idleChannel.Reader;
+            _idleConnectorWriter = idleChannel.Writer;
+
+            //TODO: initiate idle lifetime and pruning fields
         }
 
         /// <inheritdoc/>
-        internal override ValueTask<SqlConnector> GetInternalConnection(SqlConnectionX owningConnection, TimeSpan timeout, bool async, CancellationToken cancellationToken)
+        internal override async ValueTask<SqlConnector> GetInternalConnection(SqlConnectionX owningConnection, TimeSpan timeout, bool async, CancellationToken cancellationToken)
         {
-            throw new NotImplementedException();
+            CheckDisposed();
+
+            if (TryGetIdleConnector(out SqlConnector? connector))
+            {
+                return connector;
+            }
+
+            // First, try to open a new physical connector. This will fail if we're at max capacity.
+            connector = await OpenNewInternalConnection(owningConnection, timeout, async, cancellationToken).ConfigureAwait(false);
+            if (connector != null)
+            {
+                return connector;
+            }
+
+            // We're at max capacity. Block on the idle channel with a timeout.
+            // Note that Channels guarantee fair FIFO behavior to callers of ReadAsync (first-come first-
+            // served), which is crucial to us.
+            using CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            CancellationToken finalToken = linkedSource.Token;
+            linkedSource.CancelAfter(timeout);
+            //TODO: respect remaining time, linkedSource.CancelAfter(timeout.CheckAndGetTimeLeft());
+            //TODO: MetricsReporter.ReportPendingConnectionRequestStart();
```

Comment:
What is `MetricsReporter` ?

Reply:
I believe this may have been copied over from NpgSql driver. Clean up the commented out code to say something to the effect that this needs to be plumbed into the existing connection pool counters in SqlClient.

Reply:
cleaned up